File: solutions/ruby/allergies/2/allergies.rb
class Allergies
  ALLERGENS = {
    'eggs':           1,
    'peanuts':        2,
    'shellfish':      4,
    'strawberries':   8,
    'tomatoes':      16,
    'chocolate':     32,
    'pollen':        64,
    'cats':         128
  }.freeze

  def initialize(score)
    @score = score
  end

  def allergic_to?(allergen)
    allergen_flag = ALLERGENS.fetch(allergen.to_sym)
    @score & allergen_flag == allergen_flag
  end

  def list
    ALLERGENS.keys
      .map(&:to_s)
      .select { |k| allergic_to?(k) }
  end
end

File: solutions/c/allergies/2/allergies.c
#include "allergies.h"

bool is_allergic_to(allergen_t allergen, unsigned score)
{
    return (1 << allergen) & score;
}

allergen_list_t get_allergens(unsigned score)
{
    allergen_list_t list = {0};

    for (int i = 0; i < ALLERGEN_COUNT; i++)
    {
        if (is_allergic_to(i, score))
        {
            list.allergens[i] = true;
            list.count++;
        }
    }

    return list;
}
